In [163]:
import numpy as np
import pandas as pd
import pathlib as fl
import os as os
from PIL import Image

labs = pd.read_csv('labels.csv')
routes = pd.read_csv('routes.csv')

In [164]:
labs = labs.drop(labels = ['Unnamed: 0'], axis = 1)
routes = routes.drop(labels = ['Unnamed: 0'], axis = 1)

In [165]:
routes['dup'] = routes.duplicated(subset = ['1x0', '1x1', '1x2', '1x3', '1x4', '1x5', '1x6', '1x7', '1x8', '1x9', '1x10', '2x0', '2x1', '2x2', '2x3', '2x4', '2x5', '2x6', '2x7', '2x8', '2x9', '2x10', '3x0', '3x1', '3x2', '3x3', '3x4', '3x5', '3x6', '3x7', '3x8', '3x9', '3x10', '4x0', '4x1', '4x2', '4x3', '4x4', '4x5', '4x6', '4x7', '4x8', '4x9', '4x10', '5x0', '5x1', '5x2', '5x3', '5x4', '5x5', '5x6', '5x7', '5x8', '5x9', '5x10', '6x0', '6x1', '6x2', '6x3', '6x4', '6x5', '6x6', '6x7', '6x8', '6x9', '6x10', '7x0', '7x1', '7x2', '7x3', '7x4', '7x5', '7x6', '7x7', '7x8', '7x9', '7x10', '8x0', '8x1', '8x2', '8x3', '8x4', '8x5', '8x6', '8x7', '8x8', '8x9', '8x10', '9x0', '9x1', '9x2', '9x3', '9x4', '9x5', '9x6', '9x7', '9x8', '9x9', '9x10', '10x0', '10x1', '10x2', '10x3', '10x4', '10x5', '10x6', '10x7', '10x8', '10x9', '10x10', '11x0', '11x1', '11x2', '11x3', '11x4', '11x5', '11x6', '11x7', '11x8', '11x9', '11x10', '12x0', '12x1', '12x2', '12x3', '12x4', '12x5', '12x6', '12x7', '12x8', '12x9', '12x10', '13x0', '13x1', '13x2', '13x3', '13x4', '13x5', '13x6', '13x7', '13x8', '13x9', '13x10', '14x0', '14x1', '14x2', '14x3', '14x4', '14x5', '14x6', '14x7', '14x8', '14x9', '14x10', '15x0', '15x1', '15x2', '15x3', '15x4', '15x5', '15x6', '15x7', '15x8', '15x9', '15x10', '16x0', '16x1', '16x2', '16x3', '16x4', '16x5', '16x6', '16x7', '16x8', '16x9', '16x10', '17x0', '17x1', '17x2', '17x3', '17x4', '17x5', '17x6', '17x7', '17x8', '17x9', '17x10', '18x0', '18x1', '18x2', '18x3', '18x4', '18x5', '18x6', '18x7', '18x8', '18x9', '18x10'], keep = False)

In [166]:
routes.index = routes['Name']
routes = routes.drop(labels =['Name'], axis = 1)
routes

1x0 1x1 1x2 1x3 1x4 1x5 1x6 1x7 1x8 1x9  ... 18x2 18x3  \
Name                                                         ...             
# E-Z WARMUP          N   N   N   N   N   N   N   N   N   N  ...    N    N   
##1                   N   N   N   N   N   N   N   N   N   N  ...    N    N   
##2                   N   N   N   N   N   N   N   N   N   N  ...    N    N   
##3                   N   N   N   N   N   N   N   N   N   N  ...    N    N   
##4                   N   N   N   N   N   N   N   N   N   N  ...    N    N   
...                  ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ...  ...   
ZZZ BY E.             N   N   N   N   N   N   N   N   N   N  ...    F    N   
ZZZZZZZZZZ            N   N   N   N   N   N   N   N   N   N  ...    N    F   
] V5-6                N   N   N   N   N   N   N   N   N   N  ...    N    N   
_TEST1                N   N   N   N   N   N   N   N   N   N  ...    N    N   
“SKY AND SAND “   N   N   N   N   N   N   N   N   N   N  ...    N    F   

                    18x4 18x5 18x6 18x7 18x8 18x9 18x10    dup  
Name                                                            
# E-Z WARMUP           N    N    N    N    F    N     N  False  
##1                    F    N    N    N    N    N     N  False  
##2                    N    N    N    N    F    N     N  False  
##3                    N    N    N    N    N    N     F  False  
##4                    N    N    N    N    F    N     N  False  
...                  ...  ...  ...  ...  ...  ...   ...    ...  
ZZZ BY E.              N    N    N    N    N    N     N  False  
ZZZZZZZZZZ             N    N    N    N    N    N     N   True  
] V5-6                 N    N    F    N    N    N     N  False  
_TEST1                 N    N    F    N    N    N     N  False  
“SKY AND SAND “    N    N    N    N    N    N     N   True  

[29752 rows x 199 columns]

In [167]:
labs = labs.join(routes, on = 'Name',how = 'left')

In [168]:
labs['dup'] = labs.duplicated(subset = ['1x0', '1x1', '1x2', '1x3', '1x4', '1x5', '1x6', '1x7', '1x8', '1x9', '1x10', '2x0', '2x1', '2x2', '2x3', '2x4', '2x5', '2x6', '2x7', '2x8', '2x9', '2x10', '3x0', '3x1', '3x2', '3x3', '3x4', '3x5', '3x6', '3x7', '3x8', '3x9', '3x10', '4x0', '4x1', '4x2', '4x3', '4x4', '4x5', '4x6', '4x7', '4x8', '4x9', '4x10', '5x0', '5x1', '5x2', '5x3', '5x4', '5x5', '5x6', '5x7', '5x8', '5x9', '5x10', '6x0', '6x1', '6x2', '6x3', '6x4', '6x5', '6x6', '6x7', '6x8', '6x9', '6x10', '7x0', '7x1', '7x2', '7x3', '7x4', '7x5', '7x6', '7x7', '7x8', '7x9', '7x10', '8x0', '8x1', '8x2', '8x3', '8x4', '8x5', '8x6', '8x7', '8x8', '8x9', '8x10', '9x0', '9x1', '9x2', '9x3', '9x4', '9x5', '9x6', '9x7', '9x8', '9x9', '9x10', '10x0', '10x1', '10x2', '10x3', '10x4', '10x5', '10x6', '10x7', '10x8', '10x9', '10x10', '11x0', '11x1', '11x2', '11x3', '11x4', '11x5', '11x6', '11x7', '11x8', '11x9', '11x10', '12x0', '12x1', '12x2', '12x3', '12x4', '12x5', '12x6', '12x7', '12x8', '12x9', '12x10', '13x0', '13x1', '13x2', '13x3', '13x4', '13x5', '13x6', '13x7', '13x8', '13x9', '13x10', '14x0', '14x1', '14x2', '14x3', '14x4', '14x5', '14x6', '14x7', '14x8', '14x9', '14x10', '15x0', '15x1', '15x2', '15x3', '15x4', '15x5', '15x6', '15x7', '15x8', '15x9', '15x10', '16x0', '16x1', '16x2', '16x3', '16x4', '16x5', '16x6', '16x7', '16x8', '16x9', '16x10', '17x0', '17x1', '17x2', '17x3', '17x4', '17x5', '17x6', '17x7', '17x8', '17x9', '17x10', '18x0', '18x1', '18x2', '18x3', '18x4', '18x5', '18x6', '18x7', '18x8', '18x9', '18x10'])

In [169]:
# Cut to relavent data:

smaller = labs.drop(labels = ['Stars1', 'Stars2', 'Type', 'dup'], axis = 1)

In [170]:
smaller = smaller.dropna()

In [171]:
smaller = smaller.drop_duplicates(subset = ['1x0', '1x1', '1x2', '1x3', '1x4', '1x5', '1x6', '1x7', '1x8', '1x9', '1x10', '2x0', '2x1', '2x2', '2x3', '2x4', '2x5', '2x6', '2x7', '2x8', '2x9', '2x10', '3x0', '3x1', '3x2', '3x3', '3x4', '3x5', '3x6', '3x7', '3x8', '3x9', '3x10', '4x0', '4x1', '4x2', '4x3', '4x4', '4x5', '4x6', '4x7', '4x8', '4x9', '4x10', '5x0', '5x1', '5x2', '5x3', '5x4', '5x5', '5x6', '5x7', '5x8', '5x9', '5x10', '6x0', '6x1', '6x2', '6x3', '6x4', '6x5', '6x6', '6x7', '6x8', '6x9', '6x10', '7x0', '7x1', '7x2', '7x3', '7x4', '7x5', '7x6', '7x7', '7x8', '7x9', '7x10', '8x0', '8x1', '8x2', '8x3', '8x4', '8x5', '8x6', '8x7', '8x8', '8x9', '8x10', '9x0', '9x1', '9x2', '9x3', '9x4', '9x5', '9x6', '9x7', '9x8', '9x9', '9x10', '10x0', '10x1', '10x2', '10x3', '10x4', '10x5', '10x6', '10x7', '10x8', '10x9', '10x10', '11x0', '11x1', '11x2', '11x3', '11x4', '11x5', '11x6', '11x7', '11x8', '11x9', '11x10', '12x0', '12x1', '12x2', '12x3', '12x4', '12x5', '12x6', '12x7', '12x8', '12x9', '12x10', '13x0', '13x1', '13x2', '13x3', '13x4', '13x5', '13x6', '13x7', '13x8', '13x9', '13x10', '14x0', '14x1', '14x2', '14x3', '14x4', '14x5', '14x6', '14x7', '14x8', '14x9', '14x10', '15x0', '15x1', '15x2', '15x3', '15x4', '15x5', '15x6', '15x7', '15x8', '15x9', '15x10', '16x0', '16x1', '16x2', '16x3', '16x4', '16x5', '16x6', '16x7', '16x8', '16x9', '16x10', '17x0', '17x1', '17x2', '17x3', '17x4', '17x5', '17x6', '17x7', '17x8', '17x9', '17x10', '18x0', '18x1', '18x2', '18x3', '18x4', '18x5', '18x6', '18x7', '18x8', '18x9', '18x10'])

In [172]:
smaller = smaller[smaller['Repeats'] != 'Be the first to repeat this problem'] # drop low repeats
smaller['Repeats']

0                              20928 repeats
1                              29496 repeats
2                                612 repeats
3                               7494 repeats
4                                221 repeats
                        ...                 
34788                              3 repeats
34789    1 climber has repeated this problem
34791    1 climber has repeated this problem
34792                               4repeats
34794                             25 repeats
Name: Repeats, Length: 18943, dtype: object

In [173]:
vals = smaller['Repeats'].values

for i in range(len(vals)):
    vals[i] = vals[i].rstrip('repeats')
    vals[i] = vals[i].rstrip('climber has repeated this problem')
    vals[i] = vals[i].replace('S','5')
    

vals

In [174]:
print(vals)

['20928' '29496' '612' ... '1' '4' '25']


In [175]:
smaller['Repeats'] = pd.to_numeric(smaller['Repeats'], errors = 'coerce')

In [176]:
grades = smaller['Grade'].values
grades

array(['6B+ (User grade 6B+)', '6B+ (User grade 6B+)',
       '6B+ (User grade 6B+)', ..., '7C (User grade 7C)', '6C',
       '6C (User grade 6C)'], dtype=object)

In [177]:
for i in range(len(grades)):
    grades[i] = (grades[i].split('(')[0].replace('TA+', '7A+').replace('TAt', '7a+').replace('baT','6a+').replace('++','+').strip())
    if grades[i].isupper():
        pass
    else:
        grades[i] = grades[i].swapcase()

In [178]:
'7a+'.swapcase()

'7A+'

In [179]:
def valid_grade(grade):
    return grade in ['6B+', '7A', '7B', '7A+', '6C+', '7B+', '6C', '7C', '7C+','6B', '8A', '8B', '8A+', '8B+']

In [180]:
smaller

Name        Creator  Repeats Grade 1x0 1x1 1x2  \
0      FAR FROM THE MADDING CROWD       Ben Moon  20928.0   6B+   N   N   N   
1               WUTHERING HEIGHTS       Ben Moon  29496.0   6B+   N   N   N   
2                       PROBLEM 3   Ally Patrick    612.0   6B+   N   N   N   
3                      HARD TIMES       Ben Moon   7494.0    7A   N   N   N   
4                       PROBLEM 5   Ally Patrick    221.0    7A   N   N   N   
...                           ...            ...      ...   ...  ..  ..  ..   
34788                ELECTRON TOM       MTikusis      3.0    6C   N   N   N   
34789                       ATTIK       MTikusis      1.0   6B+   N   N   N   
34791               CORE CLIMBERS       Ivan Luo      1.0    7C   N   N   N   
34792      FOR THE TALL AND OLD 1  martin nygren      4.0    6C   N   N   N   
34794                   AIRPLANES    Julia Bobak     25.0    6C   N   N   N   

      1x3 1x4 1x5  ... 18x1 18x2 18x3 18x4 18x5 18x6 18x7 18x8 18x9 18x10  
0       N   N   N  ...    N    N    F    N    N    N    N    N    N     N  
1       N   N   N  ...    N    N    N    F    N    N    N    N    N     N  
2       N   N   N  ...    N    N    N    N    N    N    N    F    N     N  
3       N   N   N  ...    N    N    N    N    N    N    N    F    N     N  
4       N   N   N  ...    N    N    F    N    N    N    N    N    N     N  
...    ..  ..  ..  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...  
34788   N   N   N  ...    N    N    F    N    N    N    N    N    N     N  
34789   N   N   N  ...    N    N    N    N    N    F    N    N    N     N  
34791   N   N   N  ...    N    N    N    N    N    N    N    N    N     F  
34792   N   N   N  ...    N    N    N    N    N    F    N    N    N     N  
34794   N   N   N  ...    F    N    N    N    N    N    N    N    N     N  

[18943 rows x 202 columns]

In [181]:
smaller = smaller[smaller['Repeats'] > 3]

In [182]:
smaller

Name        Creator  Repeats Grade 1x0 1x1 1x2  \
0      FAR FROM THE MADDING CROWD       Ben Moon  20928.0   6B+   N   N   N   
1               WUTHERING HEIGHTS       Ben Moon  29496.0   6B+   N   N   N   
2                       PROBLEM 3   Ally Patrick    612.0   6B+   N   N   N   
3                      HARD TIMES       Ben Moon   7494.0    7A   N   N   N   
4                       PROBLEM 5   Ally Patrick    221.0    7A   N   N   N   
...                           ...            ...      ...   ...  ..  ..  ..   
34774              TI VOLEVO BENE       Piffer97     17.0   6B+   N   N   N   
34775          TI VOLEVO BENE VAR       Piffer97      6.0    6C   N   N   N   
34779                     PIG BIM     Joao Horst     16.0   6B+   N   N   N   
34792      FOR THE TALL AND OLD 1  martin nygren      4.0    6C   N   N   N   
34794                   AIRPLANES    Julia Bobak     25.0    6C   N   N   N   

      1x3 1x4 1x5  ... 18x1 18x2 18x3 18x4 18x5 18x6 18x7 18x8 18x9 18x10  
0       N   N   N  ...    N    N    F    N    N    N    N    N    N     N  
1       N   N   N  ...    N    N    N    F    N    N    N    N    N     N  
2       N   N   N  ...    N    N    N    N    N    N    N    F    N     N  
3       N   N   N  ...    N    N    N    N    N    N    N    F    N     N  
4       N   N   N  ...    N    N    F    N    N    N    N    N    N     N  
...    ..  ..  ..  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...  
34774   N   N   N  ...    N    N    F    N    N    N    N    N    N     N  
34775   N   N   N  ...    N    N    F    N    N    N    N    N    N     N  
34779   N   N   N  ...    N    N    F    N    N    N    N    N    N     N  
34792   N   N   N  ...    N    N    N    N    N    F    N    N    N     N  
34794   N   N   N  ...    F    N    N    N    N    N    N    N    N     N  

[11803 rows x 202 columns]

In [183]:
data = smaller.drop(labels = ['Creator', 'Name'], axis = 1)


In [184]:
grades = ['6B+', '7A', '7B', '7A+', '6C+', '7B+', '6C', '7C', '7C+','6B', '8A', '8B', '8A+', '8B+']
valid = []
for grade in data['Grade']:
    if grade in grades:
        valid.append(True)
    else:
        valid.append(False)

num_v = np.array(valid)

In [185]:
grades.sort()
grades

['6B',
 '6B+',
 '6C',
 '6C+',
 '7A',
 '7A+',
 '7B',
 '7B+',
 '7C',
 '7C+',
 '8A',
 '8A+',
 '8B',
 '8B+']

In [186]:
data = data[num_v]

In [187]:
data

Repeats Grade 1x0 1x1 1x2 1x3 1x4 1x5 1x6 1x7  ... 18x1 18x2 18x3 18x4  \
0      20928.0   6B+   N   N   N   N   N   N   N   N  ...    N    N    F    N   
1      29496.0   6B+   N   N   N   N   N   N   N   N  ...    N    N    N    F   
2        612.0   6B+   N   N   N   N   N   N   N   N  ...    N    N    N    N   
3       7494.0    7A   N   N   N   N   N   N   N   N  ...    N    N    N    N   
4        221.0    7A   N   N   N   N   N   N   N   N  ...    N    N    F    N   
...        ...   ...  ..  ..  ..  ..  ..  ..  ..  ..  ...  ...  ...  ...  ...   
34774     17.0   6B+   N   N   N   N   N   N   N   N  ...    N    N    F    N   
34775      6.0    6C   N   N   N   N   N   N   N   N  ...    N    N    F    N   
34779     16.0   6B+   N   N   N   N   N   N   N   N  ...    N    N    F    N   
34792      4.0    6C   N   N   N   N   N   N   N   N  ...    N    N    N    N   
34794     25.0    6C   N   N   N   N   N   N   N   N  ...    F    N    N    N   

      18x5 18x6 18x7 18x8 18x9 18x10  
0        N    N    N    N    N     N  
1        N    N    N    N    N     N  
2        N    N    N    F    N     N  
3        N    N    N    F    N     N  
4        N    N    N    N    N     N  
...    ...  ...  ...  ...  ...   ...  
34774    N    N    N    N    N     N  
34775    N    N    N    N    N     N  
34779    N    N    N    N    N     N  
34792    N    F    N    N    N     N  
34794    N    N    N    N    N     N  

[11803 rows x 200 columns]

In [188]:


import keras
import keras.optimizers
from keras.datasets import mnist
from keras.layers import Dense
from keras.models import Sequential
from keras.layers import Layer
from matplotlib import pyplot as plt
from tensorflow.keras.utils import to_categorical
from keras.utils.np_utils import to_categorical # convert to one-hot-encoding
from keras.models import Sequential
from keras.layers import Dense, Dropout, Flatten, Conv2D, MaxPool2D, BatchNormalization, Softmax
from keras.preprocessing.image import ImageDataGenerator
from keras.callbacks import ReduceLROnPlateau
from keras.utils.np_utils import to_categorical
from keras.losses import Loss
from tensorflow.keras.optimizers import SGD, Adam,RMSprop, Adadelta
from random import randint
import tensorflow as tf



In [189]:
from sklearn.model_selection import train_test_split
X = data.drop(labels = ['Grade', 'Repeats'], axis = 1)
Y = data['Grade']
weights = data['Repeats']
#split dataset into train and test data


In [190]:
Y

0        6B+
1        6B+
2        6B+
3         7A
4         7A
        ... 
34774    6B+
34775     6C
34779    6B+
34792     6C
34794     6C
Name: Grade, Length: 11803, dtype: object

In [191]:
weights = np.sqrt(weights)
weights = np.ceil(weights)
weights

0        145.0
1        172.0
2         25.0
3         87.0
4         15.0
         ...  
34774      5.0
34775      3.0
34779      4.0
34792      2.0
34794      5.0
Name: Repeats, Length: 11803, dtype: float64

In [192]:
X

1x0 1x1 1x2 1x3 1x4 1x5 1x6 1x7 1x8 1x9  ... 18x1 18x2 18x3 18x4 18x5  \
0       N   N   N   N   N   N   N   N   N   N  ...    N    N    F    N    N   
1       N   N   N   N   N   N   N   N   N   N  ...    N    N    N    F    N   
2       N   N   N   N   N   N   N   N   N   N  ...    N    N    N    N    N   
3       N   N   N   N   N   N   N   N   N   N  ...    N    N    N    N    N   
4       N   N   N   N   N   N   N   N   N   N  ...    N    N    F    N    N   
...    ..  ..  ..  ..  ..  ..  ..  ..  ..  ..  ...  ...  ...  ...  ...  ...   
34774   N   N   N   N   N   N   N   N   N   N  ...    N    N    F    N    N   
34775   N   N   N   N   N   N   N   N   N   N  ...    N    N    F    N    N   
34779   N   N   N   N   N   N   N   N   N   N  ...    N    N    F    N    N   
34792   N   N   N   N   N   N   N   N   N   N  ...    N    N    N    N    N   
34794   N   N   N   N   N   N   N   N   N   N  ...    F    N    N    N    N   

      18x6 18x7 18x8 18x9 18x10  
0        N    N    N    N     N  
1        N    N    N    N     N  
2        N    N    F    N     N  
3        N    N    F    N     N  
4        N    N    N    N     N  
...    ...  ...  ...  ...   ...  
34774    N    N    N    N     N  
34775    N    N    N    N     N  
34779    N    N    N    N     N  
34792    F    N    N    N     N  
34794    N    N    N    N     N  

[11803 rows x 198 columns]

In [193]:
rep_arr = []
for i in range(len(grades)):
    rep_arr.append(i)

In [194]:
Y.replace(grades, rep_arr, inplace = True)
Y

0        1
1        1
2        1
3        4
4        4
        ..
34774    1
34775    2
34779    1
34792    2
34794    2
Name: Grade, Length: 11803, dtype: int64

In [195]:
Y.unique()

array([ 1,  4,  6,  5,  3,  7,  2,  8,  9,  0, 10, 12, 11, 13],
      dtype=int64)

In [196]:
grades

['6B',
 '6B+',
 '6C',
 '6C+',
 '7A',
 '7A+',
 '7B',
 '7B+',
 '7C',
 '7C+',
 '8A',
 '8A+',
 '8B',
 '8B+']

In [197]:
# Convert X to numeric:

for col in X:
    X[col].replace(['N','U','S','F'], [0,1,1,1], inplace = True)
    

In [198]:
X

1x0  1x1  1x2  1x3  1x4  1x5  1x6  1x7  1x8  1x9  ...  18x1  18x2  \
0        0    0    0    0    0    0    0    0    0    0  ...     0     0   
1        0    0    0    0    0    0    0    0    0    0  ...     0     0   
2        0    0    0    0    0    0    0    0    0    0  ...     0     0   
3        0    0    0    0    0    0    0    0    0    0  ...     0     0   
4        0    0    0    0    0    0    0    0    0    0  ...     0     0   
...    ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   ...   ...   
34774    0    0    0    0    0    0    0    0    0    0  ...     0     0   
34775    0    0    0    0    0    0    0    0    0    0  ...     0     0   
34779    0    0    0    0    0    0    0    0    0    0  ...     0     0   
34792    0    0    0    0    0    0    0    0    0    0  ...     0     0   
34794    0    0    0    0    0    0    0    0    0    0  ...     1     0   

       18x3  18x4  18x5  18x6  18x7  18x8  18x9  18x10  
0         1     0     0     0     0     0     0      0  
1         0     1     0     0     0     0     0      0  
2         0     0     0     0     0     1     0      0  
3         0     0     0     0     0     1     0      0  
4         1     0     0     0     0     0     0      0  
...     ...   ...   ...   ...   ...   ...   ...    ...  
34774     1     0     0     0     0     0     0      0  
34775     1     0     0     0     0     0     0      0  
34779     1     0     0     0     0     0     0      0  
34792     0     0     0     1     0     0     0      0  
34794     0     0     0     0     0     0     0      0  

[11803 rows x 198 columns]

In [199]:
X_train, X_test, Y_train, Y_test, w_train, w_test = train_test_split(X, Y, weights, test_size=0.2)

## Loss Function

I don't love the loss functions that tensorflow comes with on its own. So I'm going to create my own. 

In [38]:
class GradeLoss(Loss):

  def call(self, y_true, y_pred):
    y_pred = tf.cast(y_pred, tf.float32) # Some categorical number (we use one-hot encoding)
    y_true = tf.cast(y_true, tf.int64) # Another categorical
    cur_acc = 0
    cur_acc = tf.cast(cur_acc, tf.float32)
    temp = 0
    div = tf.cast(len(y_pred), tf.float32)
    for t in range(len(y_pred)):
        temp = tf.math.square(tf.math.add(tf.math.abs(tf.math.subtract(tf.math.argmax(y_pred[t]), tf.math.reduce_max(y_true[t]))),1))
        temp = tf.cast(temp, tf.float32)
        cur_acc = tf.math.add(cur_acc, temp)
    return tf.math.divide(cur_acc, div)


In [39]:
def GradeMetric(y_true, y_pred):
    y_pred = tf.cast(y_pred, tf.float32) # Some categorical number (we use one-hot encoding)
    y_true = tf.cast(y_true, y_pred.dtype) # Another categorical
    return tf.reduce_mean(tf.math.reciprocal_no_nan(tf.math.square(tf.math.add(tf.math.abs(tf.math.subtract(y_pred,y_true)),1))))

### Explanation for our functions. 

The Grade Loss is just the difference of squares function between the two numbers. I may include an offset to penalize a differnce of one more, but I think this is pretty close. 

Our GradeMetric is the reciprocal of the difference of squares of the grades, plus one. We take this approach instead of a traditional "Categorized Correctly" approach since our categories are more like discrete numbers. There is a lot of similarity in real life between a 6B and a 6B+ (in fact, we would rank them the same in America). I may fiddle with the details more later, but giving a 0.5 for being off by one grade doesn't seem too inaccurate. Maybe I'll lower than to 0.333 or 0.25, but we'll see. 

In [40]:
inputs = keras.Input(shape=(198,)) # We give a 198 length vector

dense = Dense(64, activation="relu")
x = dense(inputs)
x = Dense(32, activation="elu")(x)
outputs = Dense(1)(x)
model = keras.Model(inputs=inputs, outputs=outputs, name="route_grader")
model.summary()
# A 198 -> 64 -> 32 -> 14 learner


Model: "route_grader"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 input_1 (InputLayer)        [(None, 198)]             0         
                                                                 
 dense (Dense)               (None, 64)                12736     
                                                                 
 dense_1 (Dense)             (None, 32)                2080      
                                                                 
 dense_2 (Dense)             (None, 1)                 33        
                                                                 
Total params: 14,849
Trainable params: 14,849
Non-trainable params: 0
_________________________________________________________________


In [119]:

epochs = 10
batch_size = 128
# lr - learning rate, momentum - additional parameter influencing gradient descent
optimizer = SGD(lr=0.001, momentum = 0.9) 
#optimizer = Adadelta()

inputs = keras.Input(shape=(198,)) # We give a 198 length vector

dense = Dense(64, activation="relu")
x = dense(inputs)
x = Dense(32, activation="elu")(x)
outputs = Dense(14)(x)
model = keras.Model(inputs=inputs, outputs=outputs, name="route_grader")

# Compile the model, asking also to keep track of accuracy at different iterations
model.compile(optimizer = optimizer, loss = 'SparseCategoricalCrossentropy', metrics = ['SparseCategoricalAccuracy'])



# Fit the model. If you want to see the progress, set verbose=True (it is also True by default)
history = model.fit(X_train, Y_train, epochs = epochs, verbose=True, shuffle=True,
                    validation_data = (X_test,Y_test), batch_size=batch_size)


Epoch 1/10


C:\Users\isaac\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\keras\optimizer_v2\gradient_descent.py:102: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(SGD, self).__init__(name, **kwargs)


74/74 [==============================] - 1s 4ms/step - loss: 2.8732 - sparse_categorical_accuracy: 0.4012 - val_loss: 2.2459 - val_sparse_categorical_accuracy: 0.4320
Epoch 2/10
74/74 [==============================] - 0s 2ms/step - loss: 2.2153 - sparse_categorical_accuracy: 0.4169 - val_loss: 2.2410 - val_sparse_categorical_accuracy: 0.4367
Epoch 3/10
74/74 [==============================] - 0s 2ms/step - loss: 2.1328 - sparse_categorical_accuracy: 0.4198 - val_loss: 2.1191 - val_sparse_categorical_accuracy: 0.4371
Epoch 4/10
74/74 [==============================] - 0s 2ms/step - loss: 2.0478 - sparse_categorical_accuracy: 0.4231 - val_loss: 2.0114 - val_sparse_categorical_accuracy: 0.4405
Epoch 5/10
74/74 [==============================] - 0s 2ms/step - loss: 2.0318 - sparse_categorical_accuracy: 0.4338 - val_loss: 2.0441 - val_sparse_categorical_accuracy: 0.4528
Epoch 6/10
74/74 [==============================] - 0s 2ms/step - loss: 1.9708 - sparse_categorical_accuracy: 0.4352 - va

In [120]:
model.save("cur_model")

INFO:tensorflow:Assets written to: cur_model\assets


In [121]:
#model.compile(optimizer = optimizer, loss = keras.losses.SparseCategoricalCrossentropy(), metrics = [newGradeMetric])
model.evaluate(X_test,Y_test,batch_size=128)

19/19 [==============================] - 0s 992us/step - loss: 1.9296 - sparse_categorical_accuracy: 0.4727


[1.9295830726623535, 0.47268107533454895]

In [122]:
 a= model.predict(X_test, batch_size = len(X_test))

In [123]:
a[0]

array([-0.36136317,  0.83159465,  0.65437347,  0.48612645,  0.63423103,
        0.0476352 ,  0.15685758,  0.08074622, -0.24818665, -0.16889384,
       -0.3431445 , -0.30807185, -0.4033618 , -0.2327342 ], dtype=float32)

In [124]:
def newGradeMetric(y_true, y_pred):
    y_pred = tf.cast(y_pred, tf.float32) # Some categorical number (we use one-hot encoding)
    y_true = tf.cast(y_true, tf.int64) # Another categorical
    cur_acc = 0
    cur_acc = tf.cast(cur_acc, tf.float32)
    temp = 0
    div = tf.cast(len(y_pred), tf.float32)
    for t in range(len(y_pred)):
        temp = tf.math.abs(tf.math.subtract(tf.math.argmax(y_pred[t]), tf.math.reduce_max(y_true[t])))
        temp = tf.experimental.numpy.power(2, temp)
        temp = tf.cast(temp, tf.float32)
        temp = tf.math.reciprocal_no_nan(temp)
        cur_acc = tf.math.add(cur_acc, temp)
    return tf.math.divide(cur_acc, div)

In [125]:
b = newGradeMetric(Y_test, a)
b

<tf.Tensor: shape=(), dtype=float32, numpy=0.6320556>

# Modifying the Model

## Sigmoid

In [215]:

epochs = 10
batch_size = 128
# lr - learning rate, momentum - additional parameter influencing gradient descent
optimizer = SGD(lr=0.001, momentum = 0.9) 
#optimizer = Adadelta()

sig_inputs = keras.Input(shape=(198,)) # We give a 198 length vector

dense = Dense(64, activation="sigmoid")
sig_x = dense(sig_inputs)
#sig_x = Dense(32, activation="sigmoid")(sig_x)
sig_outputs = Dense(14)(sig_x)
sig_model = keras.Model(inputs=sig_inputs, outputs=sig_outputs, name="route_grader")

# Compile the model, asking also to keep track of accuracy at different iterations
sig_model.compile(optimizer = optimizer, loss = 'SparseCategoricalCrossentropy', metrics = ['SparseCategoricalAccuracy'])



# Fit the model. If you want to see the progress, set verbose=True (it is also True by default)
sig_history = sig_model.fit(X_train, Y_train, epochs = epochs, verbose=True, shuffle=True,
                    validation_data = (X_test,Y_test), batch_size=batch_size)


Epoch 1/10


C:\Users\isaac\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\keras\optimizer_v2\gradient_descent.py:102: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(SGD, self).__init__(name, **kwargs)


74/74 [==============================] - 1s 3ms/step - loss: 5.8795 - sparse_categorical_accuracy: 0.3801 - val_loss: 5.9701 - val_sparse_categorical_accuracy: 0.4176
Epoch 2/10
74/74 [==============================] - 0s 1ms/step - loss: 5.7492 - sparse_categorical_accuracy: 0.4193 - val_loss: 5.9824 - val_sparse_categorical_accuracy: 0.4176
Epoch 3/10
74/74 [==============================] - 0s 2ms/step - loss: 5.7549 - sparse_categorical_accuracy: 0.4193 - val_loss: 6.0193 - val_sparse_categorical_accuracy: 0.4176
Epoch 4/10
74/74 [==============================] - 0s 2ms/step - loss: 5.7425 - sparse_categorical_accuracy: 0.4193 - val_loss: 5.9478 - val_sparse_categorical_accuracy: 0.4176
Epoch 5/10
74/74 [==============================] - 0s 1ms/step - loss: 5.7272 - sparse_categorical_accuracy: 0.4193 - val_loss: 5.9474 - val_sparse_categorical_accuracy: 0.4176
Epoch 6/10
74/74 [==============================] - 0s 2ms/step - loss: 5.7293 - sparse_categorical_accuracy: 0.4193 - va

# Continuous predictor?

What if we treat out outcome variable as a continuous value? So we just want to get as close to the true grade as possible?

In [138]:
epochs = 30
batch_size = 128
# lr - learning rate, momentum - additional parameter influencing gradient descent
coptimizer = SGD(lr=0.001, momentum = 0.9) 
#optimizer = Adadelta()

cinputs = keras.Input(shape=(198,)) # We give a 198 length vector

cdense = Dense(64, activation="relu")
cx = dense(cinputs)
cx = Dense(32, activation="elu")(cx)
coutputs = Dense(1)(cx)
c_model = keras.Model(inputs=cinputs, outputs=coutputs, name="route_grader")

# Compile the model, asking also to keep track of accuracy at different iterations
c_model.compile(optimizer = coptimizer, loss = 'mean_absolute_error', metrics = ['mean_absolute_error'])



# Fit the model. If you want to see the progress, set verbose=True (it is also True by default)
c_history = c_model.fit(X_train, Y_train, epochs = epochs, verbose=True, shuffle=True,
                    validation_data = (X_test,Y_test), batch_size=batch_size)


Epoch 1/30


C:\Users\isaac\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\keras\optimizer_v2\gradient_descent.py:102: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(SGD, self).__init__(name, **kwargs)


74/74 [==============================] - 1s 3ms/step - loss: 2.0206 - mean_absolute_error: 2.0206 - val_loss: 1.4034 - val_mean_absolute_error: 1.4034
Epoch 2/30
74/74 [==============================] - 0s 1ms/step - loss: 1.3337 - mean_absolute_error: 1.3337 - val_loss: 1.2165 - val_mean_absolute_error: 1.2165
Epoch 3/30
74/74 [==============================] - 0s 1ms/step - loss: 1.1772 - mean_absolute_error: 1.1772 - val_loss: 1.0804 - val_mean_absolute_error: 1.0804
Epoch 4/30
74/74 [==============================] - 0s 1ms/step - loss: 1.0526 - mean_absolute_error: 1.0526 - val_loss: 0.9946 - val_mean_absolute_error: 0.9946
Epoch 5/30
74/74 [==============================] - 0s 1ms/step - loss: 0.9873 - mean_absolute_error: 0.9873 - val_loss: 0.9597 - val_mean_absolute_error: 0.9597
Epoch 6/30
74/74 [==============================] - 0s 1ms/step - loss: 0.9465 - mean_absolute_error: 0.9465 - val_loss: 0.9384 - val_mean_absolute_error: 0.9384
Epoch 7/30
74/74 [=====================

In [139]:
c = c_model.predict(X_test, batch_size = len(X_test))

In [140]:
def absGradeMetric(y_true, y_pred):
    close = 1
    y_pred = tf.cast(y_pred, tf.float32) # Some categorical number (we use one-hot encoding)
    y_true = tf.cast(y_true, tf.float32) # Another categorical
    cur_acc = 0
    cur_acc = tf.cast(cur_acc, tf.float32)
    temp = 0
    div = tf.cast(len(y_pred), tf.float32)
    for t in range(len(y_pred)):
        temp = tf.math.subtract(y_true[t], y_pred[t])
        temp = tf.math.abs(temp)
        temp = tf.math.less_equal(temp, [close])
        temp = tf.math.count_nonzero(temp)
        temp = tf.cast(temp, tf.float32)
        cur_acc = tf.math.add(cur_acc, temp)
    return tf.math.divide(cur_acc, div)

In [141]:
d = absGradeMetric(Y_test, c)
d

<tf.Tensor: shape=(), dtype=float32, numpy=0.6721728>

This is pretty good! We're within 1 of our desired grade 67% of the time! 

# Adding in a Softmax (binary classification) layer:


In [150]:
epochs = 1000
batch_size = 128
# lr - learning rate, momentum - additional parameter influencing gradient descent
optimizer = SGD(lr=0.0001, momentum = 0.9) 
#optimizer = Adadelta()

sinputs = keras.Input(shape=(198,)) # We give a 198 length vector

sdense = Dense(64, activation="relu")
sx = dense(sinputs)
sx = Dense(32, activation="elu")(sx)
soutputs = Softmax()(sx)
s_model = keras.Model(inputs=sinputs, outputs=soutputs, name="route_grader")

# Compile the model, asking also to keep track of accuracy at different iterations
s_model.compile(optimizer = optimizer, loss = 'SparseCategoricalCrossentropy', metrics = [tf.keras.metrics.SparseTopKCategoricalAccuracy(k=2)])



# Fit the model. If you want to see the progress, set verbose=True (it is also True by default)
s_history = s_model.fit(X_train, Y_train, epochs = epochs, verbose=True, shuffle=True,
                    validation_data = (X_test,Y_test), batch_size=batch_size)

Epoch 1/1000


C:\Users\isaac\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\keras\optimizer_v2\gradient_descent.py:102: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(SGD, self).__init__(name, **kwargs)


74/74 [==============================] - 0s 3ms/step - loss: 3.6702 - sparse_top_k_categorical_accuracy: 0.0012 - val_loss: 3.6432 - val_sparse_top_k_categorical_accuracy: 8.4710e-04
Epoch 2/1000
74/74 [==============================] - 0s 1ms/step - loss: 3.5969 - sparse_top_k_categorical_accuracy: 0.0037 - val_loss: 3.5600 - val_sparse_top_k_categorical_accuracy: 0.0042
Epoch 3/1000
74/74 [==============================] - 0s 1ms/step - loss: 3.5067 - sparse_top_k_categorical_accuracy: 0.0098 - val_loss: 3.4563 - val_sparse_top_k_categorical_accuracy: 0.0131
Epoch 4/1000
74/74 [==============================] - 0s 1ms/step - loss: 3.4005 - sparse_top_k_categorical_accuracy: 0.0225 - val_loss: 3.3443 - val_sparse_top_k_categorical_accuracy: 0.0292
Epoch 5/1000
74/74 [==============================] - 0s 2ms/step - loss: 3.2926 - sparse_top_k_categorical_accuracy: 0.0513 - val_loss: 3.2341 - val_sparse_top_k_categorical_accuracy: 0.0733
Epoch 6/1000
74/74 [=============================

74/74 [==============================] - 0s 1ms/step - loss: 1.8669 - sparse_top_k_categorical_accuracy: 0.5810 - val_loss: 1.8200 - val_sparse_top_k_categorical_accuracy: 0.5883
Epoch 44/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.8579 - sparse_top_k_categorical_accuracy: 0.5814 - val_loss: 1.8114 - val_sparse_top_k_categorical_accuracy: 0.5879
Epoch 45/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.8493 - sparse_top_k_categorical_accuracy: 0.5813 - val_loss: 1.8030 - val_sparse_top_k_categorical_accuracy: 0.5883
Epoch 46/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.8410 - sparse_top_k_categorical_accuracy: 0.5818 - val_loss: 1.7949 - val_sparse_top_k_categorical_accuracy: 0.5892
Epoch 47/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.8330 - sparse_top_k_categorical_accuracy: 0.5825 - val_loss: 1.7871 - val_sparse_top_k_categorical_accuracy: 0.5896
Epoch 48/1000
74/74 [============================

74/74 [==============================] - 0s 1ms/step - loss: 1.6491 - sparse_top_k_categorical_accuracy: 0.6001 - val_loss: 1.6099 - val_sparse_top_k_categorical_accuracy: 0.6091
Epoch 86/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.6462 - sparse_top_k_categorical_accuracy: 0.6004 - val_loss: 1.6071 - val_sparse_top_k_categorical_accuracy: 0.6091
Epoch 87/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.6433 - sparse_top_k_categorical_accuracy: 0.6010 - val_loss: 1.6043 - val_sparse_top_k_categorical_accuracy: 0.6108
Epoch 88/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.6405 - sparse_top_k_categorical_accuracy: 0.6019 - val_loss: 1.6016 - val_sparse_top_k_categorical_accuracy: 0.6112
Epoch 89/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.6377 - sparse_top_k_categorical_accuracy: 0.6022 - val_loss: 1.5990 - val_sparse_top_k_categorical_accuracy: 0.6120
Epoch 90/1000
74/74 [============================

74/74 [==============================] - 0s 1ms/step - loss: 1.5613 - sparse_top_k_categorical_accuracy: 0.6115 - val_loss: 1.5265 - val_sparse_top_k_categorical_accuracy: 0.6260
Epoch 128/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.5599 - sparse_top_k_categorical_accuracy: 0.6113 - val_loss: 1.5251 - val_sparse_top_k_categorical_accuracy: 0.6269
Epoch 129/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.5584 - sparse_top_k_categorical_accuracy: 0.6119 - val_loss: 1.5238 - val_sparse_top_k_categorical_accuracy: 0.6269
Epoch 130/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.5570 - sparse_top_k_categorical_accuracy: 0.6126 - val_loss: 1.5224 - val_sparse_top_k_categorical_accuracy: 0.6277
Epoch 131/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.5555 - sparse_top_k_categorical_accuracy: 0.6126 - val_loss: 1.5211 - val_sparse_top_k_categorical_accuracy: 0.6294
Epoch 132/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.5123 - sparse_top_k_categorical_accuracy: 0.6224 - val_loss: 1.4805 - val_sparse_top_k_categorical_accuracy: 0.6370
Epoch 170/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.5114 - sparse_top_k_categorical_accuracy: 0.6231 - val_loss: 1.4797 - val_sparse_top_k_categorical_accuracy: 0.6370
Epoch 171/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.5105 - sparse_top_k_categorical_accuracy: 0.6235 - val_loss: 1.4788 - val_sparse_top_k_categorical_accuracy: 0.6370
Epoch 172/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.5096 - sparse_top_k_categorical_accuracy: 0.6233 - val_loss: 1.4780 - val_sparse_top_k_categorical_accuracy: 0.6374
Epoch 173/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.5087 - sparse_top_k_categorical_accuracy: 0.6237 - val_loss: 1.4772 - val_sparse_top_k_categorical_accuracy: 0.6374
Epoch 174/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.4796 - sparse_top_k_categorical_accuracy: 0.6306 - val_loss: 1.4502 - val_sparse_top_k_categorical_accuracy: 0.6472
Epoch 212/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4790 - sparse_top_k_categorical_accuracy: 0.6308 - val_loss: 1.4495 - val_sparse_top_k_categorical_accuracy: 0.6476
Epoch 213/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4783 - sparse_top_k_categorical_accuracy: 0.6313 - val_loss: 1.4490 - val_sparse_top_k_categorical_accuracy: 0.6468
Epoch 214/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4777 - sparse_top_k_categorical_accuracy: 0.6315 - val_loss: 1.4484 - val_sparse_top_k_categorical_accuracy: 0.6480
Epoch 215/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4770 - sparse_top_k_categorical_accuracy: 0.6319 - val_loss: 1.4478 - val_sparse_top_k_categorical_accuracy: 0.6472
Epoch 216/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.4557 - sparse_top_k_categorical_accuracy: 0.6358 - val_loss: 1.4281 - val_sparse_top_k_categorical_accuracy: 0.6501
Epoch 254/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4552 - sparse_top_k_categorical_accuracy: 0.6360 - val_loss: 1.4276 - val_sparse_top_k_categorical_accuracy: 0.6510
Epoch 255/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4547 - sparse_top_k_categorical_accuracy: 0.6362 - val_loss: 1.4272 - val_sparse_top_k_categorical_accuracy: 0.6510
Epoch 256/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4542 - sparse_top_k_categorical_accuracy: 0.6364 - val_loss: 1.4267 - val_sparse_top_k_categorical_accuracy: 0.6514
Epoch 257/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4537 - sparse_top_k_categorical_accuracy: 0.6364 - val_loss: 1.4263 - val_sparse_top_k_categorical_accuracy: 0.6510
Epoch 258/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.4371 - sparse_top_k_categorical_accuracy: 0.6415 - val_loss: 1.4111 - val_sparse_top_k_categorical_accuracy: 0.6531
Epoch 296/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4367 - sparse_top_k_categorical_accuracy: 0.6416 - val_loss: 1.4107 - val_sparse_top_k_categorical_accuracy: 0.6531
Epoch 297/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4363 - sparse_top_k_categorical_accuracy: 0.6418 - val_loss: 1.4104 - val_sparse_top_k_categorical_accuracy: 0.6531
Epoch 298/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4359 - sparse_top_k_categorical_accuracy: 0.6419 - val_loss: 1.4100 - val_sparse_top_k_categorical_accuracy: 0.6535
Epoch 299/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4355 - sparse_top_k_categorical_accuracy: 0.6422 - val_loss: 1.4096 - val_sparse_top_k_categorical_accuracy: 0.6535
Epoch 300/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.4218 - sparse_top_k_categorical_accuracy: 0.6463 - val_loss: 1.3974 - val_sparse_top_k_categorical_accuracy: 0.6557
Epoch 338/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4215 - sparse_top_k_categorical_accuracy: 0.6467 - val_loss: 1.3971 - val_sparse_top_k_categorical_accuracy: 0.6552
Epoch 339/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4212 - sparse_top_k_categorical_accuracy: 0.6467 - val_loss: 1.3968 - val_sparse_top_k_categorical_accuracy: 0.6552
Epoch 340/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4208 - sparse_top_k_categorical_accuracy: 0.6470 - val_loss: 1.3965 - val_sparse_top_k_categorical_accuracy: 0.6552
Epoch 341/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4205 - sparse_top_k_categorical_accuracy: 0.6470 - val_loss: 1.3962 - val_sparse_top_k_categorical_accuracy: 0.6548
Epoch 342/1000
74/74 [=======================

74/74 [==============================] - 0s 2ms/step - loss: 1.4090 - sparse_top_k_categorical_accuracy: 0.6515 - val_loss: 1.3860 - val_sparse_top_k_categorical_accuracy: 0.6586
Epoch 380/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4087 - sparse_top_k_categorical_accuracy: 0.6515 - val_loss: 1.3858 - val_sparse_top_k_categorical_accuracy: 0.6582
Epoch 381/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4084 - sparse_top_k_categorical_accuracy: 0.6516 - val_loss: 1.3855 - val_sparse_top_k_categorical_accuracy: 0.6582
Epoch 382/1000
74/74 [==============================] - 0s 2ms/step - loss: 1.4082 - sparse_top_k_categorical_accuracy: 0.6516 - val_loss: 1.3853 - val_sparse_top_k_categorical_accuracy: 0.6582
Epoch 383/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.4079 - sparse_top_k_categorical_accuracy: 0.6519 - val_loss: 1.3850 - val_sparse_top_k_categorical_accuracy: 0.6578
Epoch 384/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.3980 - sparse_top_k_categorical_accuracy: 0.6555 - val_loss: 1.3765 - val_sparse_top_k_categorical_accuracy: 0.6582
Epoch 422/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3978 - sparse_top_k_categorical_accuracy: 0.6553 - val_loss: 1.3763 - val_sparse_top_k_categorical_accuracy: 0.6586
Epoch 423/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3975 - sparse_top_k_categorical_accuracy: 0.6552 - val_loss: 1.3761 - val_sparse_top_k_categorical_accuracy: 0.6586
Epoch 424/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3973 - sparse_top_k_categorical_accuracy: 0.6555 - val_loss: 1.3758 - val_sparse_top_k_categorical_accuracy: 0.6590
Epoch 425/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3970 - sparse_top_k_categorical_accuracy: 0.6558 - val_loss: 1.3756 - val_sparse_top_k_categorical_accuracy: 0.6590
Epoch 426/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.3883 - sparse_top_k_categorical_accuracy: 0.6601 - val_loss: 1.3683 - val_sparse_top_k_categorical_accuracy: 0.6633
Epoch 464/1000
74/74 [==============================] - 0s 2ms/step - loss: 1.3881 - sparse_top_k_categorical_accuracy: 0.6596 - val_loss: 1.3682 - val_sparse_top_k_categorical_accuracy: 0.6629
Epoch 465/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3879 - sparse_top_k_categorical_accuracy: 0.6597 - val_loss: 1.3679 - val_sparse_top_k_categorical_accuracy: 0.6633
Epoch 466/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3877 - sparse_top_k_categorical_accuracy: 0.6599 - val_loss: 1.3678 - val_sparse_top_k_categorical_accuracy: 0.6629
Epoch 467/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3875 - sparse_top_k_categorical_accuracy: 0.6603 - val_loss: 1.3676 - val_sparse_top_k_categorical_accuracy: 0.6633
Epoch 468/1000
74/74 [=======================

74/74 [==============================] - 0s 2ms/step - loss: 1.3799 - sparse_top_k_categorical_accuracy: 0.6631 - val_loss: 1.3612 - val_sparse_top_k_categorical_accuracy: 0.6654
Epoch 506/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3796 - sparse_top_k_categorical_accuracy: 0.6630 - val_loss: 1.3611 - val_sparse_top_k_categorical_accuracy: 0.6654
Epoch 507/1000
74/74 [==============================] - 0s 2ms/step - loss: 1.3795 - sparse_top_k_categorical_accuracy: 0.6634 - val_loss: 1.3609 - val_sparse_top_k_categorical_accuracy: 0.6654
Epoch 508/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3793 - sparse_top_k_categorical_accuracy: 0.6633 - val_loss: 1.3608 - val_sparse_top_k_categorical_accuracy: 0.6662
Epoch 509/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3791 - sparse_top_k_categorical_accuracy: 0.6637 - val_loss: 1.3607 - val_sparse_top_k_categorical_accuracy: 0.6671
Epoch 510/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.3722 - sparse_top_k_categorical_accuracy: 0.6657 - val_loss: 1.3551 - val_sparse_top_k_categorical_accuracy: 0.6684
Epoch 548/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3720 - sparse_top_k_categorical_accuracy: 0.6644 - val_loss: 1.3550 - val_sparse_top_k_categorical_accuracy: 0.6679
Epoch 549/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3719 - sparse_top_k_categorical_accuracy: 0.6653 - val_loss: 1.3549 - val_sparse_top_k_categorical_accuracy: 0.6684
Epoch 550/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3717 - sparse_top_k_categorical_accuracy: 0.6647 - val_loss: 1.3547 - val_sparse_top_k_categorical_accuracy: 0.6684
Epoch 551/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3715 - sparse_top_k_categorical_accuracy: 0.6648 - val_loss: 1.3546 - val_sparse_top_k_categorical_accuracy: 0.6675
Epoch 552/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.3653 - sparse_top_k_categorical_accuracy: 0.6669 - val_loss: 1.3497 - val_sparse_top_k_categorical_accuracy: 0.6701
Epoch 590/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3652 - sparse_top_k_categorical_accuracy: 0.6673 - val_loss: 1.3496 - val_sparse_top_k_categorical_accuracy: 0.6709
Epoch 591/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3650 - sparse_top_k_categorical_accuracy: 0.6671 - val_loss: 1.3495 - val_sparse_top_k_categorical_accuracy: 0.6705
Epoch 592/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3648 - sparse_top_k_categorical_accuracy: 0.6675 - val_loss: 1.3493 - val_sparse_top_k_categorical_accuracy: 0.6701
Epoch 593/1000
74/74 [==============================] - 0s 2ms/step - loss: 1.3647 - sparse_top_k_categorical_accuracy: 0.6670 - val_loss: 1.3492 - val_sparse_top_k_categorical_accuracy: 0.6705
Epoch 594/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.3590 - sparse_top_k_categorical_accuracy: 0.6708 - val_loss: 1.3449 - val_sparse_top_k_categorical_accuracy: 0.6743
Epoch 632/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3589 - sparse_top_k_categorical_accuracy: 0.6705 - val_loss: 1.3448 - val_sparse_top_k_categorical_accuracy: 0.6743
Epoch 633/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3587 - sparse_top_k_categorical_accuracy: 0.6707 - val_loss: 1.3447 - val_sparse_top_k_categorical_accuracy: 0.6734
Epoch 634/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3586 - sparse_top_k_categorical_accuracy: 0.6709 - val_loss: 1.3446 - val_sparse_top_k_categorical_accuracy: 0.6734
Epoch 635/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3585 - sparse_top_k_categorical_accuracy: 0.6713 - val_loss: 1.3445 - val_sparse_top_k_categorical_accuracy: 0.6743
Epoch 636/1000
74/74 [=======================

74/74 [==============================] - 0s 2ms/step - loss: 1.3533 - sparse_top_k_categorical_accuracy: 0.6742 - val_loss: 1.3406 - val_sparse_top_k_categorical_accuracy: 0.6739
Epoch 674/1000
74/74 [==============================] - 0s 2ms/step - loss: 1.3531 - sparse_top_k_categorical_accuracy: 0.6746 - val_loss: 1.3405 - val_sparse_top_k_categorical_accuracy: 0.6739
Epoch 675/1000
74/74 [==============================] - 0s 2ms/step - loss: 1.3530 - sparse_top_k_categorical_accuracy: 0.6750 - val_loss: 1.3405 - val_sparse_top_k_categorical_accuracy: 0.6739
Epoch 676/1000
74/74 [==============================] - 0s 2ms/step - loss: 1.3528 - sparse_top_k_categorical_accuracy: 0.6745 - val_loss: 1.3404 - val_sparse_top_k_categorical_accuracy: 0.6743
Epoch 677/1000
74/74 [==============================] - 0s 2ms/step - loss: 1.3527 - sparse_top_k_categorical_accuracy: 0.6745 - val_loss: 1.3403 - val_sparse_top_k_categorical_accuracy: 0.6739
Epoch 678/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.3479 - sparse_top_k_categorical_accuracy: 0.6759 - val_loss: 1.3368 - val_sparse_top_k_categorical_accuracy: 0.6756
Epoch 716/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3478 - sparse_top_k_categorical_accuracy: 0.6767 - val_loss: 1.3367 - val_sparse_top_k_categorical_accuracy: 0.6756
Epoch 717/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3477 - sparse_top_k_categorical_accuracy: 0.6766 - val_loss: 1.3367 - val_sparse_top_k_categorical_accuracy: 0.6747
Epoch 718/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3475 - sparse_top_k_categorical_accuracy: 0.6767 - val_loss: 1.3365 - val_sparse_top_k_categorical_accuracy: 0.6756
Epoch 719/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3474 - sparse_top_k_categorical_accuracy: 0.6778 - val_loss: 1.3365 - val_sparse_top_k_categorical_accuracy: 0.6756
Epoch 720/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.3429 - sparse_top_k_categorical_accuracy: 0.6786 - val_loss: 1.3334 - val_sparse_top_k_categorical_accuracy: 0.6777
Epoch 758/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3428 - sparse_top_k_categorical_accuracy: 0.6787 - val_loss: 1.3333 - val_sparse_top_k_categorical_accuracy: 0.6789
Epoch 759/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3427 - sparse_top_k_categorical_accuracy: 0.6794 - val_loss: 1.3332 - val_sparse_top_k_categorical_accuracy: 0.6781
Epoch 760/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3426 - sparse_top_k_categorical_accuracy: 0.6786 - val_loss: 1.3331 - val_sparse_top_k_categorical_accuracy: 0.6781
Epoch 761/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3425 - sparse_top_k_categorical_accuracy: 0.6788 - val_loss: 1.3330 - val_sparse_top_k_categorical_accuracy: 0.6785
Epoch 762/1000
74/74 [=======================

74/74 [==============================] - 0s 2ms/step - loss: 1.3383 - sparse_top_k_categorical_accuracy: 0.6815 - val_loss: 1.3302 - val_sparse_top_k_categorical_accuracy: 0.6802
Epoch 800/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3382 - sparse_top_k_categorical_accuracy: 0.6816 - val_loss: 1.3301 - val_sparse_top_k_categorical_accuracy: 0.6811
Epoch 801/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3381 - sparse_top_k_categorical_accuracy: 0.6815 - val_loss: 1.3301 - val_sparse_top_k_categorical_accuracy: 0.6806
Epoch 802/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3380 - sparse_top_k_categorical_accuracy: 0.6815 - val_loss: 1.3300 - val_sparse_top_k_categorical_accuracy: 0.6806
Epoch 803/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3378 - sparse_top_k_categorical_accuracy: 0.6811 - val_loss: 1.3299 - val_sparse_top_k_categorical_accuracy: 0.6798
Epoch 804/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.3339 - sparse_top_k_categorical_accuracy: 0.6829 - val_loss: 1.3273 - val_sparse_top_k_categorical_accuracy: 0.6806
Epoch 842/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3338 - sparse_top_k_categorical_accuracy: 0.6825 - val_loss: 1.3272 - val_sparse_top_k_categorical_accuracy: 0.6806
Epoch 843/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3337 - sparse_top_k_categorical_accuracy: 0.6836 - val_loss: 1.3272 - val_sparse_top_k_categorical_accuracy: 0.6806
Epoch 844/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3336 - sparse_top_k_categorical_accuracy: 0.6832 - val_loss: 1.3271 - val_sparse_top_k_categorical_accuracy: 0.6806
Epoch 845/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3335 - sparse_top_k_categorical_accuracy: 0.6823 - val_loss: 1.3271 - val_sparse_top_k_categorical_accuracy: 0.6806
Epoch 846/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.3298 - sparse_top_k_categorical_accuracy: 0.6840 - val_loss: 1.3247 - val_sparse_top_k_categorical_accuracy: 0.6819
Epoch 884/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3297 - sparse_top_k_categorical_accuracy: 0.6840 - val_loss: 1.3246 - val_sparse_top_k_categorical_accuracy: 0.6823
Epoch 885/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3296 - sparse_top_k_categorical_accuracy: 0.6843 - val_loss: 1.3245 - val_sparse_top_k_categorical_accuracy: 0.6823
Epoch 886/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3295 - sparse_top_k_categorical_accuracy: 0.6845 - val_loss: 1.3245 - val_sparse_top_k_categorical_accuracy: 0.6828
Epoch 887/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3294 - sparse_top_k_categorical_accuracy: 0.6843 - val_loss: 1.3244 - val_sparse_top_k_categorical_accuracy: 0.6823
Epoch 888/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.3258 - sparse_top_k_categorical_accuracy: 0.6850 - val_loss: 1.3222 - val_sparse_top_k_categorical_accuracy: 0.6832
Epoch 926/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3258 - sparse_top_k_categorical_accuracy: 0.6854 - val_loss: 1.3221 - val_sparse_top_k_categorical_accuracy: 0.6832
Epoch 927/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3257 - sparse_top_k_categorical_accuracy: 0.6854 - val_loss: 1.3221 - val_sparse_top_k_categorical_accuracy: 0.6845
Epoch 928/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3256 - sparse_top_k_categorical_accuracy: 0.6853 - val_loss: 1.3220 - val_sparse_top_k_categorical_accuracy: 0.6840
Epoch 929/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3255 - sparse_top_k_categorical_accuracy: 0.6849 - val_loss: 1.3220 - val_sparse_top_k_categorical_accuracy: 0.6836
Epoch 930/1000
74/74 [=======================

74/74 [==============================] - 0s 1ms/step - loss: 1.3221 - sparse_top_k_categorical_accuracy: 0.6871 - val_loss: 1.3199 - val_sparse_top_k_categorical_accuracy: 0.6815
Epoch 968/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3220 - sparse_top_k_categorical_accuracy: 0.6871 - val_loss: 1.3198 - val_sparse_top_k_categorical_accuracy: 0.6819
Epoch 969/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3219 - sparse_top_k_categorical_accuracy: 0.6870 - val_loss: 1.3198 - val_sparse_top_k_categorical_accuracy: 0.6819
Epoch 970/1000
74/74 [==============================] - 0s 1ms/step - loss: 1.3218 - sparse_top_k_categorical_accuracy: 0.6874 - val_loss: 1.3197 - val_sparse_top_k_categorical_accuracy: 0.6815
Epoch 971/1000
74/74 [==============================] - 0s 2ms/step - loss: 1.3218 - sparse_top_k_categorical_accuracy: 0.6870 - val_loss: 1.3196 - val_sparse_top_k_categorical_accuracy: 0.6815
Epoch 972/1000
74/74 [=======================

In [151]:
sf = s_model.predict(X_test, batch_size = len(X_test))
sf[0]

array([1.22573139e-04, 3.80574584e-01, 2.35781670e-01, 1.97212636e-01,
       1.10107735e-01, 4.08565104e-02, 1.92431752e-02, 8.10545124e-03,
       3.70686571e-03, 1.32983772e-03, 8.07381992e-04, 1.26739193e-04,
       9.69749890e-05, 9.46169675e-05, 9.82670899e-05, 9.88695465e-05,
       1.05119900e-04, 1.15803385e-04, 9.15868732e-05, 1.00845733e-04,
       1.02387225e-04, 1.02893195e-04, 1.01839069e-04, 1.12609843e-04,
       9.79878387e-05, 9.04943081e-05, 1.00432546e-04, 1.00704456e-04,
       8.74791949e-05, 1.09521403e-04, 1.16355121e-04, 1.00018864e-04],
      dtype=float32)

## K = 1

In [152]:
s_model.compile(optimizer = optimizer, loss = 'SparseCategoricalCrossentropy', metrics = [tf.keras.metrics.SparseTopKCategoricalAccuracy(k=1)])
s_model.evaluate(X_test,Y_test, batch_size = len(X_test))

1/1 [==============================] - 0s 99ms/step - loss: 1.3181 - sparse_top_k_categorical_accuracy: 0.4956


[1.3181487321853638, 0.4955527186393738]

## K = 2

In [153]:
s_model.compile(optimizer = optimizer, loss = 'SparseCategoricalCrossentropy', metrics = [tf.keras.metrics.SparseTopKCategoricalAccuracy(k=2)])
s_model.evaluate(X_test,Y_test, batch_size = len(X_test))

1/1 [==============================] - 0s 98ms/step - loss: 1.3181 - sparse_top_k_categorical_accuracy: 0.6823


[1.3181487321853638, 0.6823379993438721]

## K = 3

In [155]:
s_model.compile(optimizer = optimizer, loss = 'SparseCategoricalCrossentropy', metrics = [tf.keras.metrics.SparseTopKCategoricalAccuracy(k=3)])
s_model.evaluate(X_test,Y_test, batch_size = len(X_test))

1/1 [==============================] - 0s 100ms/step - loss: 1.3181 - sparse_top_k_categorical_accuracy: 0.8153


[1.3181487321853638, 0.8153324723243713]

So our model will find the true grade of climbing routes within the top 3 guesses 81.5% of the time! And within the top two 68.2% of the time!

# Making The Grade

The 6A-8C scale is a little wide for my tastes. Let's try condensing the grades, and retesting our initial models to see if that helps. We'll go from 14 classes to either 9 or 10. 

In [448]:
grades

['6B',
 '6B+',
 '6C',
 '6C+',
 '7A',
 '7A+',
 '7B',
 '7B+',
 '7C',
 '7C+',
 '8A',
 '8A+',
 '8B',
 '8B+']

In [450]:
for i in range(len(rep_arr)):
    print(rep_arr[i], grades[i])

0 6B
1 6B+
2 6C
3 6C+
4 7A
5 7A+
6 7B
7 7B+
8 7C
9 7C+
10 8A
11 8A+
12 8B
13 8B+


In [156]:
Y.replace(rep_arr, [0, 0,1,1,2,3,4,4,5,6,7,8,9,10], inplace = True)


In [157]:
X_train, X_test, Y_train, Y_test, w_train, w_test = train_test_split(X, Y, weights, test_size=0.2)

In [158]:
epochs = 1000
batch_size = 128
# lr - learning rate, momentum - additional parameter influencing gradient descent
optimizer = SGD(lr=0.0001, momentum = 0.9) 
#optimizer = Adadelta()

gsinputs = keras.Input(shape=(198,)) # We give a 198 length vector

gsdense = Dense(64, activation="relu")
gsx = dense(gsinputs)
gsx = Dense(32, activation="elu")(gsx)
gsoutputs = Softmax()(gsx)
gs_model = keras.Model(inputs=gsinputs, outputs=gsoutputs, name="route_grader")

# Compile the model, asking also to keep track of accuracy at different iterations
gs_model.compile(optimizer = optimizer, loss = 'SparseCategoricalCrossentropy', metrics = [tf.keras.metrics.SparseTopKCategoricalAccuracy(k=2)])



# Fit the model. If you want to see the progress, set verbose=True (it is also True by default)
gs_history = gs_model.fit(X_train, Y_train, epochs = epochs, verbose=False, shuffle=True,
                    validation_data = (X_test,Y_test), batch_size=batch_size)

C:\Users\isaac\AppData\Local\Packages\PythonSoftwareFoundation.Python.3.9_qbz5n2kfra8p0\LocalCache\local-packages\Python39\site-packages\keras\optimizer_v2\gradient_descent.py:102: UserWarning: The `lr` argument is deprecated, use `learning_rate` instead.
  super(SGD, self).__init__(name, **kwargs)


In [159]:
gs_model.evaluate(X_test,Y_test, batch_size = len(X_test))

1/1 [==============================] - 0s 101ms/step - loss: 1.1063 - sparse_top_k_categorical_accuracy: 0.7920


[1.1062678098678589, 0.7920372486114502]

In [160]:
gs_model.compile(optimizer = optimizer, loss = 'SparseCategoricalCrossentropy', metrics = [tf.keras.metrics.SparseTopKCategoricalAccuracy(k=1)])
gs_model.evaluate(X_test,Y_test, batch_size = len(X_test))

1/1 [==============================] - 0s 108ms/step - loss: 1.1063 - sparse_top_k_categorical_accuracy: 0.5379


[1.1062678098678589, 0.5379076600074768]

In [161]:
gs_model.compile(optimizer = optimizer, loss = 'SparseCategoricalCrossentropy', metrics = [tf.keras.metrics.SparseTopKCategoricalAccuracy(k=3)])
gs_model.evaluate(X_test,Y_test, batch_size = len(X_test))

1/1 [==============================] - 0s 106ms/step - loss: 1.1063 - sparse_top_k_categorical_accuracy: 0.8831


[1.1062678098678589, 0.8831003904342651]

So one of our top 2 guesses for grade is right 80% of the time! But... our number one guess is right just 53% of the time